test/fixtures/parser/ansi/select_true_and_not_false.yml

```diff
@@ -39,7 +39,8 @@ file:
         keyword: SELECT
         select_clause_element:
           expression:
-            keyword: NOT
+            data_type:
+              data_type_identifier: NOT
```

Comment:
This seems weird. Why does it now think `NOT` is a `data_type`?

Reply:
I think we probably need to start being more explicit with our data types, it's matching the pattern Seq(datatype, literal), datatype is too open. It's probably a different feature in itself?

Reply:
It's not ideal, but I don't think it has any negative effects. DataTypeIdentifiers are broad and defined as:

```
    # Maybe data types should be more restrictive?
    DatatypeIdentifierSegment=RegexParser(
        r"[A-Z][A-Z0-9_]*",
        CodeSegment,
        name="data_type_identifier",
        type="data_type_identifier",
    ),
```

Reply:
Agree it's not really a problem, but it is something we changed in this PR and seems odd. Suggested a potential fix below.

Reply:
Changing the `DatatypeIdentifierSegment` to the below (similar to how `NakedIdentifierSegment` was defined above it) seems to fix it and still pass all test cases, but probably some more keywords we need to add here and not just `NOT`:

```py
    # Maybe data types should be more restrictive?
    # Added NOT exclusion for now - see https://github.com/sqlfluff/sqlfluff/pull/1338
    DatatypeIdentifierSegment=SegmentGenerator(
        # Generate the anti template from the set of reserved keywords
        lambda dialect: RegexParser(
            r"[A-Z][A-Z0-9_]*",
            CodeSegment,
            name="data_type_identifier",
            type="data_type_identifier",
            anti_template=r"^(NOT)$",
        ),
    ),
```

What do you think about adding this as a start to at least keep this PR from regressing the known cases?

I'm just a little nervous about sending a key keyword like `NOT` down a different path for a base ANSI test case in case there's any knock on consequences to that - which is why the YML files are there after all to catch things like this.

Reply:
Okay, I've implemented as you suggested. I've also added a TODO on it, because ,in my opinion, this we should be explicit with types, and not keywords that can't be types